components/zcash_address/src/kind/unified.rs

```diff
@@ -140,82 +143,36 @@ impl fmt::Display for ParseError {
-pub enum Receiver {
-    Orchard([u8; 43]),
-    Sapling(kind::sapling::Data),
-    P2pkh(kind::p2pkh::Data),
-    P2sh(kind::p2sh::Data),
-    Unknown { typecode: u32, data: Vec<u8> },
-}
-
-impl cmp::Ord for Receiver {
-    fn cmp(&self, other: &Self) -> cmp::Ordering {
-        match self.typecode().cmp(&other.typecode()) {
-            cmp::Ordering::Equal => self.addr().cmp(other.addr()),
-            res => res,
-        }
-    }
-}
-
-impl cmp::PartialOrd for Receiver {
-    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
-        Some(self.cmp(other))
-    }
-}
-
-impl TryFrom<(u32, &[u8])> for Receiver {
-    type Error = ParseError;
-
-    fn try_from((typecode, addr): (u32, &[u8])) -> Result<Self, Self::Error> {
-        match typecode.try_into()? {
-            Typecode::P2pkh => addr.try_into().map(Receiver::P2pkh),
-            Typecode::P2sh => addr.try_into().map(Receiver::P2sh),
-            Typecode::Sapling => addr.try_into().map(Receiver::Sapling),
-            Typecode::Orchard => addr.try_into().map(Receiver::Orchard),
-            Typecode::Unknown(_) => Ok(Receiver::Unknown {
-                typecode,
-                data: addr.to_vec(),
-            }),
-        }
-        .map_err(|e| {
-            ParseError::InvalidEncoding(format!("Invalid address for typecode {}: {}", typecode, e))
-        })
-    }
-}
+mod private {
+    use super::{ParseError, Typecode};
+    use std::{cmp, convert::TryFrom};
 
-impl Receiver {
-    fn typecode(&self) -> Typecode {
-        match self {
-            Receiver::P2pkh(_) => Typecode::P2pkh,
-            Receiver::P2sh(_) => Typecode::P2sh,
-            Receiver::Sapling(_) => Typecode::Sapling,
-            Receiver::Orchard(_) => Typecode::Orchard,
-            Receiver::Unknown { typecode, .. } => Typecode::Unknown(*typecode),
-        }
+    /// A raw address or viewing key.
+    pub trait Receiver:
+        TryFrom<(u32, Vec<u8>), Error = ParseError> + cmp::Ord + cmp::PartialOrd + Clone
+    {
+        fn typecode(&self) -> Typecode;
+        fn data(&self) -> &[u8];
     }
 
-    fn addr(&self) -> &[u8] {
-        match self {
-            Receiver::P2pkh(data) => data,
-            Receiver::P2sh(data) => data,
-            Receiver::Sapling(data) => data,
-            Receiver::Orchard(data) => data,
-            Receiver::Unknown { data, .. } => data,
-        }
+    pub trait Sealed {
+        type Receiver: Receiver;
+        fn from_inner(receivers: Vec<Self::Receiver>) -> Self;
     }
 }
 
-/// A Unified Address.
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct Address(pub(crate) Vec<Receiver>);
+use private::Receiver;
 
-impl TryFrom<(&str, &[u8])> for Address {
-    type Error = ParseError;
+/// Trait providing common encoding logic for Unified containers.
+pub trait Unified: private::Sealed + std::marker::Sized {
+    const MAINNET: &'static str;
+    const TESTNET: &'static str;
+    const REGTEST: &'static str;
```

Comment:
```suggestion
```
Given that these constants aren't used anywhere within the trait's methods, I think it's a mistake to make them constants here.